src/python/pants/backend/experimental/jvm/resolve/coursier_setup.py

```diff
@@ -0,0 +1,196 @@
+from dataclasses import dataclass
+
+from pants.core.util_rules.external_tool import (
+    DownloadedExternalTool,
+    ExternalTool,
+    ExternalToolRequest,
+)
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.platform import Platform
+from pants.engine.rules import Get, collect_rules, rule
+
+logger = logging.getLogger(__name__)
+
+
+COURSIER_POST_PROCESSING_SCRIPT = textwrap.dedent(
+    """\
+import json
+import sys
+from pathlib import PurePath
+from shutil import copyfile
+
+report = json.load(open(sys.argv[1]))
+
+classpath = set()
+for dep in report['dependencies']:
+    file_path = PurePath(dep['file'])
+    classpath_dest = f"classpath/{file_path.name}"
+    if classpath_dest in classpath:
+        raise Exception(f"Found duplicate jar name {file_path.name}, which isn't currently supported")
+    classpath.add(classpath_dest)
+    copyfile(file_path, classpath_dest)
+"""
+)
+
+
+@dataclass(frozen=True)
+class CoursierPostProcessingScript:
+    digest: Digest
+    relpath: str
+
+
+COURSIER_WRAPPER_SCRIPT = textwrap.dedent(
+    """\
+set -eux
+
+coursier_exe="$1"
+shift
+json_output_file="$1"
+shift
+
+"$coursier_exe" fetch --json-output-file="$json_output_file" "$@"
+
+/bin/mkdir -p classpath
+/usr/bin/python3 coursier_post_processing_script.py "$json_output_file"
+"""
+)
+
+JAVAC_WRAPPER_SCRIPT = textwrap.dedent(
+    """\
+set -eux
+
+coursier_exe="$1"
+shift
+
+javac_path="$(${coursier_exe} java-home --jvm openjdk:1.9.0-4)/bin/javac"
+
+exec "${javac_path}" "$@"
+"""
+)
+
+
+@dataclass(frozen=True)
+class CoursierJavac:
+    digest: Digest
+    relpath: str
+
+
+@dataclass(frozen=True)
+class CoursierWrapperScript:
+    digest: Digest
+    relpath: str
+
+
+@dataclass(frozen=True)
+class SetupCoursierRequest:
+    pass
+
+
+class CoursierBinary(ExternalTool):
+    options_scope = "coursier"
+    help = "A dependency resolver for the Maven ecosystem."
+
+    default_version = "v2.0.13"
+    default_known_versions = [
+        "v2.0.13|linux|1ae089789cc4b0a4d296d6852b760d7f8bf72805267a6b7571e99b681d5e13b4|59652208",
+        "v2.0.13|darwin|d74b8fe4ffc2f4e9011d7151722fc8b5ffca8a72b3bc4188c61df3326228c4ef|57625024",
+    ]
+
+    def generate_url(self, plat: Platform) -> str:
+        plat_str = "x86_64-pc-linux" if plat == Platform.linux else "x86_64-apple-darwin"
+        return (
+            f"https://github.com/coursier/coursier/releases/download/{self.version}/cs-{plat_str}"
+        )
+
+
+@dataclass(frozen=True)
+class Coursier:
+    """The Coursier tool and various utilities, materialzed to a `Digest` and ready to use."""
+
+    coursier: DownloadedExternalTool
+    wrapper_script: CoursierWrapperScript
+    post_processing_script: CoursierPostProcessingScript
+    javac: CoursierJavac
+    digest: Digest
+
+
+@rule
+async def setup_coursier(
+    request: SetupCoursierRequest, coursier_binary: CoursierBinary
+) -> Coursier:
+    downloaded_coursier = await Get(
```

Comment:
You have several `await Get`s in this rule, which are all blocking / synchronous. It'd be better to parallelize via `await MultiGet`.

https://www.pantsbuild.org/docs/rules-api-tips#tip-use-multiget-for-increased-concurrency

It often gets too overwhelming to have that many `Get`s inlined, so you can do something like:

```python
download_coursier_get = Get(DownloadedExternalTool, ExternalToolRequest, ...)
wrapper_script_get = Get(Digest, CreateDigest, ...)
downloaded_coursier, wrapper_script = await MultiGet(download_coursier_get, wrapper_script_get)
```

Reply:
Although..looking at the callsite of `Coursier`, I think you can simplify a lot of this? Iiuc, you don't need to store the individual `Digest` for each of these intermediate types like the wrapper script. I think you can simplify `Coursier` so that it only stores a single merged `Digest`, and then several `str` attributes with the path to things like `downloaded_coursier.exe` and the wrapper script.

If you make that change, you would only need three `Get`s:
1. Download the tool
2. `CreateDigest`, with multiple `FileContent`s in it for each of the scripts
3. `MergeDigests`

The 1st and 2nd of those can be combined into a single `MultiGet` for concurrency.

Reply:
Done